File: plugins/push-todo/scripts/daemon.py
#!/usr/bin/env python3
"""
Push Task Execution Daemon

Polls Supabase for queued tasks and executes them via Claude Code.
Auto-heals (starts) on any /push-todo command via daemon_health.py.

Architecture:
- Git branch = worktree = Claude session (1:1:1 mapping)
- Uses Claude's --continue to resume sessions in worktrees
- SessionEnd hook reports completion (no wrapper script needed)

See: /docs/20260127_parallel_task_execution_research.md
See: /docs/20260127_parallel_task_execution_implementation_plan.md
"""

import json
import os
import signal
import subprocess
import sys
import time
import urllib.request
import urllib.error
import urllib.parse
from datetime import datetime
from pathlib import Path
from typing import List, Optional, Dict, Any

# ==================== Configuration ====================

API_BASE_URL = "https://jxuzqcbqhiaxmfitzxlo.supabase.co/functions/v1"
POLL_INTERVAL = 30  # seconds
MAX_CONCURRENT_TASKS = 3  # Max parallel Claude sessions

PID_FILE = Path.home() / ".push" / "daemon.pid"
LOG_FILE = Path.home() / ".push" / "daemon.log"
CONFIG_FILE = Path.home() / ".config" / "push" / "config"

# Track running tasks to avoid duplicates
running_tasks: Dict[int, subprocess.Popen] = {}


# ==================== Logging ====================

def log(message: str):
    """Log with timestamp to both stdout and log file."""
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    line = f"[{timestamp}] {message}"
    print(line, flush=True)
    try:
        with open(LOG_FILE, "a") as f:
            f.write(line + "\n")
    except Exception:
        pass


# ==================== Configuration ====================

def get_api_key() -> Optional[str]:
    """Get API key from config file or environment."""
    # Check environment first
    if os.environ.get("PUSH_API_KEY"):
        return os.environ["PUSH_API_KEY"]

    # Then check config file
    if CONFIG_FILE.exists():
        try:
            for line in CONFIG_FILE.read_text().splitlines():
                line = line.strip()
                if line.startswith("export PUSH_API_KEY="):
                    value = line.split("=", 1)[1].strip()
                    return value.strip('"').strip("'")
        except Exception as e:
            log(f"Error reading config: {e}")
    return None


def get_git_remote() -> Optional[str]:
    """Get normalized git remote for current directory."""
    try:
        result = subprocess.run(
            ["git", "remote", "get-url", "origin"],
            capture_output=True,
            text=True,
            timeout=5
        )
        if result.returncode != 0:
            return None

        url = result.stdout.strip()

        # Normalize: remove protocol prefixes
        for prefix in ["https://", "http://", "git@", "ssh://git@"]:
            if url.startswith(prefix):
                url = url[len(prefix):]
                break

        # Normalize: git@github.com:user/repo -> github.com/user/repo
        if ":" in url and "://" not in url:
            url = url.replace(":", "/", 1)

        # Remove .git suffix
        if url.endswith(".git"):
            url = url[:-4]

        return url
    except Exception:
        return None


# ==================== API Helpers ====================

def api_request(
    endpoint: str,
    method: str = "GET",
    data: Optional[Dict] = None,
    timeout: int = 15
) -> Optional[Dict]:
    """Make authenticated API request to Supabase."""
    api_key = get_api_key()
    if not api_key:
        log("No API key configured")
        return None

    url = f"{API_BASE_URL}/{endpoint}"

    req = urllib.request.Request(url, method=method)
    req.add_header("Authorization", f"Bearer {api_key}")
    req.add_header("Content-Type", "application/json")

    if data:
        req.data = json.dumps(data).encode()

    try:
        with urllib.request.urlopen(req, timeout=timeout) as response:
            return json.loads(response.read().decode())
    except urllib.error.HTTPError as e:
        log(f"API error ({e.code}): {e.reason}")
        return None
    except Exception as e:
        log(f"Request error: {e}")
        return None


def update_task_status(
    display_number: int,
    status: str,
    summary: Optional[str] = None,
    error: Optional[str] = None
):
    """Update task execution status via edge function."""
    payload: Dict[str, Any] = {
        "displayNumber": display_number,
        "status": status,
    }
    if summary:
        payload["summary"] = summary
    if error:
        payload["error"] = error

    result = api_request("update-task-execution", method="PATCH", data=payload)
    if result and result.get("success"):
        log(f"Updated task #{display_number} to {status}")
    else:
        log(f"Failed to update task #{display_number}")


# ==================== Task Fetching ====================

def fetch_queued_tasks() -> List[Dict]:
    """Fetch tasks with execution_status='queued' from Supabase."""
    git_remote = get_git_remote()

    # Build query params
    params = {"execution_status": "queued"}
    if git_remote:
        params["git_remote"] = git_remote

    query_string = urllib.parse.urlencode(params)
    endpoint = f"synced-todos?{query_string}"

    result = api_request(endpoint)
    if result:
        return result.get("todos", [])
    return []


# ==================== Task Execution ====================

def get_worktree_path(display_number: int) -> Path:
    """Get the worktree path for a task."""
    # Worktrees go in parent of current working directory
    return Path.cwd().parent / f"push-{display_number}"


def create_worktree(display_number: int) -> bool:
    """Create git worktree for task if it doesn't exist."""
    branch = f"push-{display_number}"
    worktree_path = get_worktree_path(display_number)

    if worktree_path.exists():
        log(f"Worktree already exists: {worktree_path}")
        return True

    try:
        result = subprocess.run(
            ["git", "worktree", "add", str(worktree_path), "-b", branch],
            capture_output=True,
            text=True,
            timeout=30
        )

        if result.returncode == 0:
            log(f"Created worktree: {worktree_path}")
            return True
        else:
            # Branch might already exist, try without -b
            result = subprocess.run(
                ["git", "worktree", "add", str(worktree_path), branch],
                capture_output=True,
                text=True,
                timeout=30
            )
            if result.returncode == 0:
                log(f"Created worktree (existing branch): {worktree_path}")
                return True
            else:
                log(f"Failed to create worktree: {result.stderr}")
                return False

    except Exception as e:
        log(f"Worktree creation error: {e}")
        return False


def execute_task(task: Dict):
    """Create worktree and run Claude for a task."""
    display_num = task.get("displayNumber") or task.get("display_number")
    content = (
        task.get("normalizedContent") or
        task.get("normalized_content") or
        task.get("summary") or
        "Work on this task"
    )

    if not display_num:
        log(f"Task has no display number, skipping")
        return

    if display_num in running_tasks:
        log(f"Task #{display_num} already running, skipping")
        return

    if len(running_tasks) >= MAX_CONCURRENT_TASKS:
        log(f"Max concurrent tasks ({MAX_CONCURRENT_TASKS}) reached, skipping #{display_num}")
        return

    log(f"Starting task #{display_num}: {content[:60]}...")

    # Update status to running
    update_task_status(display_num, "running")

    # Create worktree
    if not create_worktree(display_num):
        update_task_status(display_num, "failed", error="Failed to create git worktree")
        return

    worktree_path = get_worktree_path(display_num)

    # Build prompt for Claude
    prompt = f"""Work on Push task #{display_num}:

{content}

IMPORTANT: When you're done, the SessionEnd hook will automatically report completion to Supabase.

If you need to understand the codebase, start by reading the CLAUDE.md file if it exists."""

    try:
        # Run Claude in headless mode
        # SessionEnd hook will handle reporting completion
        proc = subprocess.Popen(
            [
                "claude",
                "-p", prompt,
                "--allowedTools", "Read,Edit,Write,Glob,Grep,Bash(git *)",
                "--output-format", "json"
            ],
            cwd=str(worktree_path),
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )

        # Track the running process
        running_tasks[display_num] = proc

        log(f"Started Claude for task #{display_num} (PID: {proc.pid})")

    except Exception as e:
        log(f"Error starting Claude for task #{display_num}: {e}")
        update_task_status(display_num, "failed", error=str(e))


def check_running_tasks():
    """Check status of running tasks and clean up completed ones."""
    completed = []

    for display_num, proc in running_tasks.items():
        retcode = proc.poll()

        if retcode is not None:
            # Process completed
            completed.append(display_num)

            if retcode == 0:
                log(f"Task #{display_num} completed (Claude exited cleanly)")
                # SessionEnd hook should have already reported completion
                # But we'll set a fallback summary just in case
                # update_task_status(display_num, "completed", summary="Completed via daemon")
            else:
                log(f"Task #{display_num} failed (Claude exit code: {retcode})")
                stderr = proc.stderr.read() if proc.stderr else ""
                update_task_status(display_num, "failed", error=f"Exit code {retcode}: {stderr[:200]}")

    # Remove completed tasks from tracking
    for display_num in completed:
        del running_tasks[display_num]


# ==================== Signal Handling ====================

def cleanup(signum, frame):
    """Handle shutdown gracefully."""
    log("Daemon shutting down...")

    # Terminate any running Claude processes
    for display_num, proc in running_tasks.items():
        log(f"Terminating task #{display_num}")
        proc.terminate()

    # Remove PID file
    try:
        PID_FILE.unlink(missing_ok=True)
    except Exception:
        pass

    sys.exit(0)


# ==================== Main Loop ====================

def main():
    """Main daemon loop."""
    # Set up signal handlers
    signal.signal(signal.SIGTERM, cleanup)
    signal.signal(signal.SIGINT, cleanup)

    # Create directories
    PID_FILE.parent.mkdir(parents=True, exist_ok=True)
    LOG_FILE.parent.mkdir(parents=True, exist_ok=True)

    # Write PID file
    PID_FILE.write_text(str(os.getpid()))

    log("=" * 60)
    log("Push task execution daemon started")
    log(f"PID: {os.getpid()}")
    log(f"Polling interval: {POLL_INTERVAL}s")
    log(f"Max concurrent tasks: {MAX_CONCURRENT_TASKS}")
    log(f"Working directory: {Path.cwd()}")
    log(f"Log file: {LOG_FILE}")
    log("=" * 60)

    # Check for API key
    if not get_api_key():
        log("WARNING: No API key configured. Run '/push-todo connect' first.")

    while True:
        try:
            # Check running tasks for completion
            check_running_tasks()

            # Fetch new queued tasks
            tasks = fetch_queued_tasks()

            if tasks:
                log(f"Found {len(tasks)} queued task(s)")
                for task in tasks:
                    execute_task(task)
            elif len(running_tasks) > 0:
                log(f"No new tasks. {len(running_tasks)} task(s) running.")

        except KeyboardInterrupt:
            cleanup(None, None)
        except Exception as e:
            log(f"Error in main loop: {e}")

        time.sleep(POLL_INTERVAL)


if __name__ == "__main__":
    main()

File: npm/push-todo/lib/daemon.js
#!/usr/bin/env node
/**
 * Push Task Execution Daemon
 *
 * Polls Supabase for queued tasks and executes them via Claude Code.
 * Auto-heals (starts) on any /push-todo command via daemon-health.js.
 *
 * Architecture:
 * - Git branch = worktree = Claude session (1:1:1 mapping)
 * - Uses Claude's --continue to resume sessions in worktrees
 * - Certainty analysis determines execution mode
 *
 * Ported from: plugins/push-todo/scripts/daemon.py
 */

import { spawn, execSync } from 'child_process';
import { existsSync, readFileSync, writeFileSync, mkdirSync, appendFileSync, unlinkSync } from 'fs';
import { homedir } from 'os';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// ==================== Configuration ====================

const API_BASE_URL = 'https://jxuzqcbqhiaxmfitzxlo.supabase.co/functions/v1';
const POLL_INTERVAL = 30000; // 30 seconds
const MAX_CONCURRENT_TASKS = 5;
const TASK_TIMEOUT_MS = 3600000; // 1 hour
const RETRY_MAX_ATTEMPTS = 3;
const RETRY_INITIAL_DELAY = 2000;

// Certainty thresholds (same as Python)
const CERTAINTY_HIGH_THRESHOLD = 0.7;
const CERTAINTY_LOW_THRESHOLD = 0.4;

// Paths
const PUSH_DIR = join(homedir(), '.push');
const CONFIG_DIR = join(homedir(), '.config', 'push');
const PID_FILE = join(PUSH_DIR, 'daemon.pid');
const LOG_FILE = join(PUSH_DIR, 'daemon.log');
const STATUS_FILE = join(PUSH_DIR, 'daemon_status.json');
const VERSION_FILE = join(PUSH_DIR, 'daemon.version');
const CONFIG_FILE = join(CONFIG_DIR, 'config');

// State
const runningTasks = new Map(); // displayNumber -> { process, task, startTime }
const completedToday = [];
let startedAt = null;

// ==================== Logging ====================

function log(message, level = 'INFO') {
  const timestamp = new Date().toISOString();
  const line = `[${timestamp}] [${level}] ${message}\n`;

  if (process.env.PUSH_DAEMON !== '1') {
    process.stdout.write(line);
  }

  try {
    appendFileSync(LOG_FILE, line);
  } catch {}
}

function logError(message) {
  log(message, 'ERROR');
}

// ==================== Config ====================

function getApiKey() {
  if (process.env.PUSH_API_KEY) {
    return process.env.PUSH_API_KEY;
  }

  if (existsSync(CONFIG_FILE)) {
    try {
      const content = readFileSync(CONFIG_FILE, 'utf8');
      const match = content.match(/^export\s+PUSH_API_KEY\s*=\s*["']?([^"'\n]+)["']?/m);
      if (match) return match[1];
    } catch {}
  }

  return null;
}

function getMachineId() {
  const machineIdFile = join(CONFIG_DIR, 'machine_id');
  if (existsSync(machineIdFile)) {
    try {
      return readFileSync(machineIdFile, 'utf8').trim();
    } catch {}
  }
  return null;
}

// ==================== API ====================

async function apiRequest(endpoint, options = {}) {
  const apiKey = getApiKey();
  if (!apiKey) {
    throw new Error('No API key configured');
  }

  const url = `${API_BASE_URL}/${endpoint}`;
  const response = await fetch(url, {
    ...options,
    headers: {
      'Authorization': `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
      ...options.headers
    }
  });

  return response;
}

async function fetchQueuedTasks() {
  try {
    const machineId = getMachineId();
    const params = new URLSearchParams();
    params.set('status', 'queued');
    if (machineId) {
      params.set('machine_id', machineId);
    }

    const response = await apiRequest(`queued-tasks?${params}`);

    if (!response.ok) {
      if (response.status === 404) return [];
      throw new Error(`HTTP ${response.status}`);
    }

    const data = await response.json();
    return data.tasks || [];
  } catch (error) {
    logError(`Failed to fetch queued tasks: ${error.message}`);
    return [];
  }
}

async function updateTaskStatus(displayNumber, status, extra = {}) {
  try {
    const payload = {
      displayNumber,
      status,
      ...extra
    };

    const response = await apiRequest('update-task-execution', {
      method: 'PATCH',
      body: JSON.stringify(payload)
    });

    return response.ok;
  } catch (error) {
    logError(`Failed to update task status: ${error.message}`);
    return false;
  }
}

// ==================== Project Registry ====================

function getProjectPath(gitRemote) {
  const registryFile = join(CONFIG_DIR, 'projects.json');

  if (!existsSync(registryFile)) {
    return null;
  }

  try {
    const data = JSON.parse(readFileSync(registryFile, 'utf8'));
    return data.projects?.[gitRemote]?.path || null;
  } catch {
    return null;
  }
}

// ==================== Task Execution ====================

function executeTask(task) {
  const displayNumber = task.displayNumber || task.display_number;
  const gitRemote = task.gitRemote || task.git_remote;
  const summary = task.summary || 'No summary';

  log(`Starting task #${displayNumber}: ${summary}`);

  // Get project path
  const projectPath = getProjectPath(gitRemote);
  if (!projectPath) {
    logError(`No project path for ${gitRemote}`);
    updateTaskStatus(displayNumber, 'failed', {
      error: `Project not registered: ${gitRemote}`
    });
    return null;
  }

  if (!existsSync(projectPath)) {
    logError(`Project path does not exist: ${projectPath}`);
    updateTaskStatus(displayNumber, 'failed', {
      error: `Project path not found: ${projectPath}`
    });
    return null;
  }

  // Build prompt for Claude
  const content = task.normalizedContent || task.content || task.summary || '';
  const transcript = task.originalTranscript || task.transcript || '';

  let prompt = `Task #${displayNumber}: ${summary}\n\n`;
  if (content) {
    prompt += `${content}\n\n`;
  }
  if (transcript && transcript !== content) {
    prompt += `Original voice note: "${transcript}"\n`;
  }

  // Update status to running
  updateTaskStatus(displayNumber, 'running');

  // Spawn Claude process
  const claudeArgs = ['--print', '--dangerously-skip-permissions', '-p', prompt];

  const child = spawn('claude', claudeArgs, {
    cwd: projectPath,
    stdio: ['ignore', 'pipe', 'pipe'],
    env: {
      ...process.env,
      PUSH_TASK_ID: task.id,
      PUSH_DISPLAY_NUMBER: String(displayNumber)
    }
  });

  const taskInfo = {
    process: child,
    task,
    displayNumber,
    startTime: Date.now(),
    stdout: '',
    stderr: ''
  };

  // Capture output
  child.stdout.on('data', (data) => {
    taskInfo.stdout += data.toString();
  });

  child.stderr.on('data', (data) => {
    taskInfo.stderr += data.toString();
  });

  // Handle completion
  child.on('close', (code) => {
    runningTasks.delete(displayNumber);

    const duration = Math.floor((Date.now() - taskInfo.startTime) / 1000);
    log(`Task #${displayNumber} completed with code ${code} (${duration}s)`);

    if (code === 0) {
      updateTaskStatus(displayNumber, 'completed', {
        duration,
        output: taskInfo.stdout.slice(-1000) // Last 1000 chars
      });

      completedToday.push({
        displayNumber,
        summary,
        completedAt: new Date().toISOString(),
        duration
      });
    } else {
      updateTaskStatus(displayNumber, 'failed', {
        error: `Exit code ${code}`,
        stderr: taskInfo.stderr.slice(-500)
      });
    }

    updateStatusFile();
  });

  child.on('error', (error) => {
    runningTasks.delete(displayNumber);
    logError(`Task #${displayNumber} error: ${error.message}`);

    updateTaskStatus(displayNumber, 'failed', {
      error: error.message
    });

    updateStatusFile();
  });

  runningTasks.set(displayNumber, taskInfo);
  return taskInfo;
}

// ==================== Status File ====================

function updateStatusFile() {
  const status = {
    running: true,
    pid: process.pid,
    startedAt,
    version: getVersion(),
    runningTasks: Array.from(runningTasks.values()).map(t => ({
      displayNumber: t.displayNumber,
      summary: t.task.summary || 'No summary',
      status: 'running',
      startTime: new Date(t.startTime).toISOString()
    })),
    queuedTasks: [],
    completedToday: completedToday.slice(-20), // Last 20
    updatedAt: new Date().toISOString()
  };

  try {
    writeFileSync(STATUS_FILE, JSON.stringify(status, null, 2));
  } catch {}
}

function getVersion() {
  try {
    const pkgPath = join(__dirname, '..', 'package.json');
    const pkg = JSON.parse(readFileSync(pkgPath, 'utf8'));
    return pkg.version || '3.0.0';
  } catch {
    return '3.0.0';
  }
}

// ==================== Main Loop ====================

async function pollAndExecute() {
  // Check for available slots
  if (runningTasks.size >= MAX_CONCURRENT_TASKS) {
    log(`All ${MAX_CONCURRENT_TASKS} slots in use, skipping poll`);
    return;
  }

  const availableSlots = MAX_CONCURRENT_TASKS - runningTasks.size;

  // Fetch queued tasks
  const tasks = await fetchQueuedTasks();

  if (tasks.length === 0) {
    return;
  }

  log(`Found ${tasks.length} queued tasks, ${availableSlots} slots available`);

  // Execute tasks up to available slots
  for (const task of tasks.slice(0, availableSlots)) {
    const displayNumber = task.displayNumber || task.display_number;

    // Skip if already running
    if (runningTasks.has(displayNumber)) {
      continue;
    }

    executeTask(task);
  }

  updateStatusFile();
}

async function checkTimeouts() {
  const now = Date.now();

  for (const [displayNumber, taskInfo] of runningTasks) {
    const elapsed = now - taskInfo.startTime;

    if (elapsed > TASK_TIMEOUT_MS) {
      log(`Task #${displayNumber} timed out after ${Math.floor(elapsed / 1000)}s`);

      try {
        taskInfo.process.kill('SIGTERM');
      } catch {}

      runningTasks.delete(displayNumber);

      updateTaskStatus(displayNumber, 'failed', {
        error: 'Task timed out'
      });
    }
  }
}

async function mainLoop() {
  log('Daemon starting...');

  startedAt = new Date().toISOString();

  // Write version file
  try {
    writeFileSync(VERSION_FILE, getVersion());
  } catch {}

  // Initial status
  updateStatusFile();

  // Main poll loop
  const poll = async () => {
    try {
      await pollAndExecute();
      await checkTimeouts();
    } catch (error) {
      logError(`Poll error: ${error.message}`);
    }
  };

  // Initial poll
  await poll();

  // Set up interval
  setInterval(poll, POLL_INTERVAL);

  log(`Daemon running (PID: ${process.pid}, poll interval: ${POLL_INTERVAL / 1000}s)`);
}

// ==================== Signal Handling ====================

function cleanup() {
  log('Daemon shutting down...');

  // Kill running tasks
  for (const [displayNumber, taskInfo] of runningTasks) {
    log(`Killing task #${displayNumber}`);
    try {
      taskInfo.process.kill('SIGTERM');
    } catch {}
  }

  // Clean up files
  try { unlinkSync(PID_FILE); } catch {}

  // Update status
  try {
    writeFileSync(STATUS_FILE, JSON.stringify({
      running: false,
      stoppedAt: new Date().toISOString()
    }));
  } catch {}

  process.exit(0);
}

process.on('SIGTERM', cleanup);
process.on('SIGINT', cleanup);
process.on('uncaughtException', (error) => {
  logError(`Uncaught exception: ${error.message}`);
  cleanup();
});

// ==================== Entry Point ====================

// Ensure directories exist
mkdirSync(PUSH_DIR, { recursive: true });

// Write PID file
writeFileSync(PID_FILE, String(process.pid));

// Start main loop
mainLoop().catch((error) => {
  logError(`Fatal error: ${error.message}`);
  cleanup();
});
